services/wallet/saved_addresses.go

```diff
@@ -15,41 +19,169 @@ type SavedAddress struct {
+	_, err = insert.Exec(sa.ChainID, sa.Address, sa.Name, sa.Favourite, meta.Removed, meta.SyncClock, meta.UpdateClock)
 	return err
 }
 
-func (sam *SavedAddressesManager) DeleteSavedAddress(chainID uint64, address common.Address) error {
-	_, err := sam.db.Exec(`DELETE FROM saved_addresses WHERE address = ? AND network_id = ?`, address, chainID)
+func (sam *SavedAddressesManager) AddOrUpdateSavedAddress(sa SavedAddress) (updatedClock int64, err error) {
+	updateClock := time.Now().Unix()
+	err = sam.addRawSavedAddress(sa, SavedAddressMeta{false, syncClockCreatedEditedHere, sql.NullInt64{Int64: updateClock, Valid: true}})
+	if err != nil {
+		return 0, err
+	}
+	return updateClock, nil
+}
+
+func (sam *SavedAddressesManager) checkIfNewerChange(ChainID uint64, Address common.Address, updateClock int64) (newer bool, err error) {
+	rows, err := sam.db.Query("SELECT update_clock FROM saved_addresses WHERE network_id = ? AND address = ?", ChainID, Address)
+	if err != nil {
+		return false, err
+	}
+	defer rows.Close()
+
+	if rows.Next() {
+		var dbUpdateClock int64
+		err := rows.Scan(&dbUpdateClock)
+		if err != nil {
+			return false, err
+		}
+
+		return dbUpdateClock <= updateClock, nil
+	}
+	return true, nil
+}
+
+func (sam *SavedAddressesManager) AddSavedAddressIfNewerUpdate(sa SavedAddress, syncClock int64, updateClock int64) (insertedOrUpdated bool, err error) {
+	newer, err := sam.checkIfNewerChange(sa.ChainID, sa.Address, updateClock)
+	if !newer {
+		return false, err
+	}
+
+	err = sam.addRawSavedAddress(sa, SavedAddressMeta{false, sql.NullInt64{Int64: syncClock, Valid: true}, sql.NullInt64{Int64: updateClock, Valid: true}})
+	if err != nil {
+		return false, err
+	}
+
+	return true, err
+}
+
+func (sam *SavedAddressesManager) DeleteSavedAddressIfNewerUpdate(chainID uint64, address common.Address, syncClock int64, updateClock int64) (deleted bool, err error) {
+	newer, err := sam.checkIfNewerChange(chainID, address, updateClock)
+	if !newer {
+		return false, err
+	}
+
+	insert, err := sam.db.Prepare(`INSERT OR REPLACE INTO saved_addresses (network_id, address, name, favourite, removed, sync_clock, update_clock) VALUES (?, ?, "", 0, 1, ?, ?)`)
+	if err != nil {
+		return false, err
+	}
+	_, err = insert.Exec(chainID, address, syncClock, updateClock)
+	if err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+func (sam *SavedAddressesManager) DeleteSavedAddress(chainID uint64, address common.Address) (updatedClock int64, err error) {
+	insert, err := sam.db.Prepare(`INSERT OR REPLACE INTO saved_addresses (network_id, address, name, favourite, removed, sync_clock, update_clock) VALUES (?, ?, "", 0, 1, NULL, ?)`)
+	if err != nil {
+		return 0, err
+	}
+	updateClock := time.Now().Unix()
+	_, err = insert.Exec(chainID, address, updateClock)
+	if err != nil {
+		return 0, err
+	}
+	return updateClock, nil
+}
+
+func (sam *SavedAddressesManager) DeleteSoftRemovedSavedAddresses(threshold int64) error {
+	tx, err := sam.db.Begin()
```

Comment:
No need to use a transaction here, unless you are doing multiple operation, but  this is not it seems